sdk/azidentity/chained_token_credential.go

```diff
@@ -36,32 +43,52 @@ func NewChainedTokenCredential(sources []azcore.TokenCredential, options *Chaine
 	}
 	cp := make([]azcore.TokenCredential, len(sources))
 	copy(cp, sources)
-	return &ChainedTokenCredential{sources: cp}, nil
+	credentialOptions := ChainedTokenCredentialOptions{}
+	if options != nil {
+		credentialOptions = *options
+	}
+	return &ChainedTokenCredential{sources: cp, retryAllSources: credentialOptions.RetryAllSources}, nil
 }
 
 // GetToken calls GetToken on the chained credentials in turn, stopping when one returns a token. This method is called automatically by Azure SDK clients.
 // ctx: Context controlling the request lifetime.
 // opts: Options for the token request, in particular the desired scope of the access token.
 func (c *ChainedTokenCredential) GetToken(ctx context.Context, opts policy.TokenRequestOptions) (token *azcore.AccessToken, err error) {
 	var errList []CredentialUnavailableError
+
+	formatError := func(err error) error {
+		var authFailed AuthenticationFailedError
+		if errors.As(err, &authFailed) {
+			err = fmt.Errorf("Authentication failed:\n%s\n%s"+createChainedErrorMessage(errList), err)
+			authErr := newAuthenticationFailedError(err, authFailed.RawResponse())
+			return authErr
+		}
+		return err
+	}
+
+	if c.successfulCredential != nil {
+		token, err = c.successfulCredential.GetToken(ctx, opts)
+		if err != nil {
```

Comment:
I think this block needs to be rewritten to handle the `retryAllSources == true` case yes?

Reply:
I don’t think so! As long as `successfulCredential` is never set, this part will never be used. In my current approach I’m not setting `successfulCredential` when `retryAllSources` is set, see here: https://github.com/Azure/azure-sdk-for-go/pull/16392/files#diff-d1472b3315b6cb3473bb118e734a540711129ee1b51d4f3f89b5eca7fd2c4295R87

Reply:
Of course, this could be changed so that `successfulCredential` is always set, but the conditional on line 69 is assumed to be false if `retryAllSources` is set. What makes the most sense?

Reply:
I get it now.  The way I was thinking about this is that we'd always use `successfulCredential`, and if it were to suddenly start failing on future requests, we'd fall back to try the others if `retryAllSources` is `true`.  The way you have it now preserves the old behavior which is probably better anyways in case anybody depens on that.

Reply:
Doing the change!

Reply:
Pushed an update here: https://github.com/Azure/azure-sdk-for-go/pull/16392/commits/a1bbdd1629549e699082252aefd84aa5376c3f83